Improve SpinBox interaction, split arrows, add theme attributes

davthedev opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->
Discussion here: https://github.com/godotengine/godot-proposals/discussions/9260

This improves the usability & styleability of the SpinBox control.

- Separate up and down buttons, each one having its own icon
- Interaction feedback when hovered, pressed and disabled arrow buttons
- When the value is at the extremity of the allowed range, one of the arrows is disabled
- Stylable arrow buttons & icons, separators, configurable minimum buttons width & spacing with the edit field. Each button support different styles, different backgrounds, icons & tinting (modulate) colors can be set for each interaction state
- The embedded LineEdit receives a theme variation name so that it is easy to style it a different way from the regular LineEdit using a global theme

Care is taken not to break the default theme & existing projects that depend on it.
- Slightly larger default width of the arrow buttons, enough to have a better usability with the mouse without risking breaking themes too much
- The single icon for both arrows is still supported; an empty icon is supplied for it in the proposed default theme change
- The behaviour to size the buttons width according to the button icons always works, although it is now possible to disable it in the theme properties.

![spinbox_01](https://github.com/godotengine/godot-proposals/assets/26961646/2cd68fc8-b0e3-4562-8dec-a7f0c5b74e22)

With some styling, what is possible. This resembles what I want to achieve in my project:

![spinbox_02](https://github.com/godotengine/godot-proposals/assets/26961646/3919bd51-219c-4936-ab56-35fc213d0554)

Big buttons for touch screens

![spinbox_03](https://github.com/godotengine/godot-proposals/assets/26961646/3b20a3c8-32c5-415a-abeb-27a29c7d6199)

Another possible minimal look with the new separator theme option

![spinbox_04](https://github.com/godotengine/godot-proposals/assets/26961646/7eadf8ee-a903-45ba-b354-b99c7d5463d8)

Live in Godot in the color selector

![spinbox_05](https://github.com/godotengine/godot-proposals/assets/26961646/86d44dd7-350f-4790-a034-532e0eee4d00)

Actually, there are two different widgets that render this UI element. The second one is a specialized cell of Tree, and seems to be used inside the inspector and project / editor options to set the various values. This PR only addresses the SpinBox standalone control. To see the result on Godot itself, use the color selector popup anywhere.

davthedev committed:
Improve SpinBox interaction, split arrows, add theme attributes

KoBeWi commented:
Perhaps `updown` could be deprecated.

MewPurPur reviewed (approved):
If I'm to believe that the old code is correct, then the new code should also be correct.

Mickeon commented:
All of these are straight upgrades but I'm just ever so slightly concerned by _how many_ theme items have been added by this single PR

MewPurPur commented:
IMO This amount of configurability should be the rule for Godot's interactive GUI, not the exception.

Mickeon commented:
It's moreso the fact that some are not even utilized by the editor itself, which I feel like should be the main point of reference to base all of this customization from. For example, the `field_and_buttons_separator` and `up_down_buttons_separator` styleboxes..

MewPurPur commented:
Okay, that's a fair point... I disagree about basing GUI customization off of the Godot editor, but I agree that these two theme items may be a bit overkill.

davthedev commented:
The configurability seems a little verbose, I agree. IMHO this is not a problem to have multiple theme attributes. They remain hidden in their own submenu when you do not actively edit those and drawing performance is not affected.

Let's not forget that Godot came from being a Nintendo DS game authoring tool AFAIK. There has been a long way since. I am proposing update to core widgets to add proper support for interaction feedback; the thing that gives that snappy feeling and improves usability.

The pattern to set styles for normal/hover/pressed/disabled/etc... states of a given customizable resource is pretty common to ensure customization of interaction feedback. Those multiple combinations of states and resources make up the large number of theme resources added. 

I thought about something to address this verbosity when having those combinations of resources and states, that would be proposed separately. Would be state-aware theme resources for example, like ColorStateLists of the early days of Android, in which one single slot receives the resources that defines the customization for the desired states instead of having as many slots as there are different states. Would allow combining states together, defining only a few of those and inherit the rest automatically... For example, you would be able to define only a normal and pressed style and automatically inherit the normal style on other states instead of having to fill them all.
Such a system would be a major change and would need to wait for a major Godot version upgrade where breaking changes can be added. 

Regarding the separators, I think each graphic used to render parts of a control should be exposed to theming. For example, there are separators in ItemList but you can only set a color; not a graphic - something to fix in a different PR.

The solution I propose here is usable right now and does not break current projects.

KoBeWi commented:
> When the value is at the extremity of the allowed range, one of the arrows is disabled

When both are at extremity, both arrows should be disabled:

![image](https://github.com/godotengine/godot/assets/2223172/7d7031fb-3a37-47be-93b1-17d027907512)

It works correctly at runtime though.

KoBeWi commented on doc/classes/SpinBox.xml:
```diff
@@ -71,8 +71,98 @@
+			Width of the horizontal separation between the text input field ([LineEdit]) and the buttons.
+		</theme_item>
+		<theme_item name="set_min_buttons_width_from_icons" data_type="constant" type="int" default="1">
+			If not [code]0[/code], the minimum button width corresponds to the widest of all icons set on those buttons, even if [theme_item buttons_width] is smaller.
```
Unless there is use-case for negative width, `buttons_width` being `-1` could be interpreted as using width from icons.

davthedev replied:
These are the possible cases:
- Do not enforce min width, just use the icon for that
- Enforce min width and let the icon enlarge it
- Enforce width and do not let the icon enlarge the buttons - even if there is some overflow. Useful for dense admin panels for example.

In the updated default & editor themes, I set the min width to have a little more of clickable area for the buttons. At the same time, existing themes may rely on the icon autogrow feature and I do not want to break it. So I configured those to use the second case.

KoBeWi replied:
What I mean is that it's redundant to have both `buttons_width` and `set_min_buttons_width_from_icons`. If you use the latter, the width is useless. Unless it's bigger than button width, in which case width from icons is useless.

We can assume `buttons_width` being negative to have the same effect as `set_min_buttons_width_from_icons` being enabled. Because either you set the width, or let it be determined automatically. This allows to remove `set_min_buttons_width_from_icons`. Unless there is a case I'm missing?

Mickeon replied:
I'd like elaboration on this. What KoBeWi is saying makes sense to me, as well.

KoBeWi commented on scene/gui/spin_box.cpp:
```diff
@@ -397,13 +547,50 @@ void SpinBox::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "custom_arrow_step", PROPERTY_HINT_RANGE, "0,10000,0.0001,or_greater"), "set_custom_arrow_step", "get_custom_arrow_step");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "select_all_on_focus"), "set_select_all_on_focus", "is_select_all_on_focus");
 
+	BIND_THEME_ITEM_CUSTOM(Theme::DATA_TYPE_CONSTANT, SpinBox, buttons_v_separation, "buttons_vertical_separation");
```
That looks like overusing `BIND_THEME_ITEM_CUSTOM`. It should only be used when cache entry does not match the item name and, unless it improves clarity, they should always match.

davthedev replied:
I still have to understand those binding macros better. I favor human readability in displayed property names in the editor, is having just "v" acceptable instead of "vertical"?

davthedev replied:
Need to get more familiar with those BIND_THEME_ macros.

Is it acceptable if the user sees "v" instead of "vertical"? I put the priority in human readability initially.

KoBeWi commented on editor/themes/editor_theme_manager.cpp:
```diff
@@ -1451,8 +1451,44 @@ void EditorThemeManager::_populate_standard_styles(const Ref<EditorTheme> &p_the
-	p_theme->set_icon("updown_disabled", "SpinBox", p_theme->get_icon(SNAME("GuiSpinboxUpdownDisabled"), EditorStringName(EditorIcons)));
+	{
+		Ref<Texture2D> empty_icon = memnew(ImageTexture);
+		p_theme->set_icon("updown", "SpinBox", empty_icon);
```
Is setting empty icon needed for anything? At worst it will be null, but that can be (and should be) handled in code.

davthedev replied:
So far, I noticed that the generated empty icon was used elsewhere in the theme. If I don't do this, the value is inherited from the parent theme. I noticed the technique is used elsewhere, so I took my inspiration from it. Nothing new about generating empty icons in the editor theme.
For this specific case, I want to avoid breaking existing themes that rely on the combined up/down icon, while setting it to empty by default.

davthedev replied:
If I don't set anything, it inherits from the parent theme instead. In this specific case, I want to avoid breaking the existing projects that may still have customized the combined up/down icon. In the updated theme, as there are the individual arrows now available, I want to get rid of the combined up/down.
There should be a way to say "nothing", distinct from the default "inherit" behavior", without having to create an empty resource just for that, I agree.

This empty icon technique is not invented by me; I spotted it elsewhere in the theme and copied it over.

KoBeWi commented on scene/gui/spin_box.cpp:
```diff
@@ -137,25 +139,51 @@ void SpinBox::_release_mouse() {
+	bool mouse_on_up_button = false;
+	bool mouse_on_down_button = false;
+	if (mb.is_valid() || mm.is_valid()) {
+		Rect2 up_button_rc = Rect2(sizing_cache.buttons_left, sizing_cache.buttons_top, sizing_cache.buttons_width, sizing_cache.button_up_height);
```
Why these are 4 separate values instead of caching it as Rect2?

davthedev replied:
I would say this is a style detail and I have not seen any consensus on it so far. Some values are shared between both button rectangles and used in calculations, such as buttons_width.

KoBeWi replied:
Then maybe you could use Vector2 instead? (for position and width) There is a Rect2 constructor from vectors.

davthedev commented:
> So I looked at the list of added properties and I agree that there is too many of them. Configurability is nice, but do we even need **that much**? Every up and down icon has its stylebox equivalent, so you can basically use styleboxes instead of icons. That's overkill. I'd just add a single stylebox for up/down background. 

With a single background resource used for both the up and down buttons, I would not be able to achieve the styles in my examples. The up button has a corner on the top, while the down button has a corner on the bottom, for example.

So far there is no "stacked" stylebox available. Therefore, in order to draw the arrow icon and the bg rectangle, with the arrow alignment performed by the control, it would require a stylebox and a texture.
Being able to draw both the arrow and the entire rectangle of the button is important for some styles and usability.

There is consistency with other buttons in toolbars where you can set the icon and the bg stylebox separately.

It is also much easier to tweak colors if you have the arrow icon and the bg stylebox set as separate elements.
Making the same in a one-stylebox operation would only work with StyleBoxTexture, defeating easy tweaking of colors of a StyleBoxRect without providing a new graphic. Customizing the editor theme is a scenario where easy tweaking matters.

> `set_min_buttons_width_from_icons` can be replaced with `buttons_width` as I mentioned. Bool constants are awkward and negative width makes the buttons go outside the SpinBox, which might be useful hack, but it can break layout. 

Yup, there are no native bool types for theme items, I have seen the constant hack used elsewhere and it is not straightforward to understand it at first sight. Just replicated the technique.

This is for a case where you want, at the same time, to have a minimum width AND let the icon grow it larger if necessary. Because, in the proposed change to the default and editor theme, I gave a minimum width to the buttons, so that they are convenient to target with the mouse, and not break the current themes in projects that may rely on the icon alone to widen up the buttons. 

Alternatively, it is possible to disable this icon-based-width behavior at all and be able to have smaller buttons than the icon would dictate. Cases where the icon has some white space on the edges or you need to have densely-packed controls in some admin window can benefit of this. 

> I'm not sure about modulates (although in Control context they should be called `color`), they can be _kind of_ useful if you want to use a single icon, but aside from fast prototyping, there is little advantage over simply using colored icons.
> 

If you only want to change the color and not the graphic, can be more straightforward. Tinting icons is used extensively in controls, sometimes with hardcoded values that are not exposed to the theme, for example for the disabled state.

> That's what I think at least, the properties are still up to discussion. Keep in mind that it's easier to add a property than to delete it later.

Sure, and I need to make sure not to break existing stuff. More thorough refactorings have to wait a major version upgrade.
As I mentioned earlier in the discussion, this significant number of theme properties can be reduced by my idea of state-aware theme resources. For example, you may be able to set a single resource for the icon color, and it would manage on its own the normal/hover/pressed/etc... states without bloating up the properties list.

KoBeWi commented on scene/gui/spin_box.cpp:
```diff
@@ -397,13 +547,50 @@ void SpinBox::_bind_methods() {
 }
 
 SpinBox::SpinBox() {
+	connect("mouse_exited", callable_mp(this, &SpinBox::_mouse_exited));
```
You can use `NOTIFICATION_MOUSE_EXIT` instead.

KoBeWi commented on scene/gui/spin_box.cpp:
```diff
@@ -239,41 +285,128 @@ void SpinBox::_line_edit_focus_exit() {
+	sizing_cache.buttons_width = w - theme_cache.field_and_buttons_separation;
+	sizing_cache.buttons_v_separation = CLAMP(theme_cache.buttons_v_separation, 0, size.height);
+	sizing_cache.buttons_left = is_layout_rtl() ? 0 : size.width - sizing_cache.buttons_width;
+	sizing_cache.buttons_top = 0;
```
Looks like this is always 0. You could remove it I guess.

KoBeWi commented on scene/gui/spin_box.cpp:
```diff
@@ -239,41 +285,128 @@ void SpinBox::_line_edit_focus_exit() {
+			Color up_icon_modulate = theme_cache.up_icon_modulate;
+			Color down_icon_modulate = theme_cache.down_icon_modulate;
+
+			bool is_fully_disabled = !line_edit->is_editable();
```
```suggestion
			bool is_fully_disabled = !is_editable();
```
Better IMO.

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/f6a78f83aa4b74aa5cb80ca2e3419448b1998e4f), it works as expected. This is a nice UX improvement :slightly_smiling_face:

https://github.com/godotengine/godot/assets/180032/b7c8adcd-da34-40e5-b30f-3ef14f942c1d

In the editor with a light theme (theme switching works in the editor without requiring a restart):

![image](https://github.com/godotengine/godot/assets/180032/c722597e-9e27-46e5-849c-e89a6dc972a9)

akien-mga commented:
@davthedev Will you be able to do the style/code quality changes suggested by @KoBeWi?

AThousandShips commented on scene/gui/spin_box.cpp:
```diff
@@ -166,7 +194,9 @@ void SpinBox::gui_input(const Ref<InputEvent> &p_event) {
 				line_edit->grab_focus();
-				set_value((up ? get_max() : get_min()));
+				if (mouse_on_up_button || mouse_on_down_button) {
+					set_value((mouse_on_up_button ? get_max() : get_min()));
```
```suggestion
					set_value(mouse_on_up_button ? get_max() : get_min());
```

davthedev commented:
Sure, I'll shortly get back to it.

I had to care about my health during the past weeks and had no energy to put on Godot dev for a while; nearly recovered now.

davthedev commented:
Rebased and code style changes are in - I believe this should be good to merge
@KoBeWi @akien-mga

KoBeWi commented:
You didn't address all my comments though.

https://github.com/godotengine/godot/pull/89265#issuecomment-2004233077 is still relevant.
Also https://github.com/godotengine/godot/pull/89265#discussion_r1528803125

KoBeWi commented on scene/gui/spin_box.cpp:
```diff
@@ -32,11 +32,13 @@
 
 #include "core/input/input.h"
 #include "core/math/expression.h"
+//#include "core/math/vector2i.h"
+//#include "scene/resources/texture.h"
```
To remove.

KoBeWi commented:
Seems like this breaks EditorSpinSliders. Integer properties no longer show arrows (e.g. `frame_coords` in Sprite2D):
![image](https://github.com/user-attachments/assets/fe16bc06-97d6-4d85-b2f7-967d0364baea)
Correct:
![image](https://github.com/user-attachments/assets/08fff0c6-1baa-4de3-8d17-008ef56b6a2b)

davthedev commented:
Oops, thanks for pointing out the problem in EditorSpinSliders. Need fix before merge.

For the other comments, I already checked and:

Both arrows are greyed out when min and max are the same when running the project; works correctly. But I did not notice that the issue was when the control is rendered in edit mode. I have not worked with edit-time-specifics yet.

The presence of the "enlarge with icon size" option is explained above as well. To sum it up, here we have 3 cases:
- Fixed size when size is a fixed number, independent of the image width
- Size adjusted to image width when size is -1
- Whichever is greater between the two, similar as if you put a min-width in CSS (when the said option is enabled and size is a fixed number)

I have put a greater minimum button width than the icon itself in the editor & default themes to improve usability. For this purpose, I could have gone with the fixed size option. But this is going to break all projects in which the icon is set to a bigger one. This is why I added the third case.

We can remove the "enlarge with icon size" option, but we have to agree with having this possible project-breaking situation in this case.

KoBeWi commented:
> I have put a greater minimum button width than the icon itself in the editor & default themes to improve usability.

Ah, I guess that justifies it.
Though it means this property exists only for the sake of compatibility. It should be somehow marked as such (e.g. with a comment, or `#ifndef DISABLE_DEPRECATED`). I think you can implement the alternate negative behavior anyway, so if the property is removed in the future, the functionality is preserved.

davthedev commented:
What do you mean by implementing the alternate negative behavior?

AThousandShips commented on scene/gui/spin_box.cpp:
```diff
@@ -239,41 +283,131 @@ void SpinBox::_line_edit_focus_exit() {
+
+	int w = theme_cache.set_min_buttons_width_from_icons != 0 ? MAX(buttons_block_icon_enforced_width, buttons_block_wanted_width) : buttons_block_wanted_width;
+
+	if ((w != sizing_cache.buttons_block_width)) {
```
```suggestion
	if (w != sizing_cache.buttons_block_width) {
```

davthedev commented:
For the missing arrows in the EditorSpinSliders, found the culprit. The EditorSpinSlider directly references theme elements from the SpinBox instead of having its own theme variables. I need to make adjustments on that side before we can merge.

davthedev commented:
The missing arrows in EditorSpinSliders problem is now fixed; OK for merging on my side.

KoBeWi commented on doc/classes/EditorSpinSlider.xml:
```diff
@@ -51,4 +51,12 @@
 	</signals>
+	<theme_items>
+		<theme_item name="updown" data_type="icon" type="Texture2D">
+			Single up/down graphic for both up/down buttons. Likely to be deprecated when individual up/down buttons will be implemented.
```
I'm not sure about that deprecation note.

Mickeon replied:
No, yeah, if it's not deprecated **now** there is no point.

If I may suggest, but it's a nitpick:
```suggestion
			Single texture representing both the up and down buttons.
```
or
```suggestion
			Single texture combining both the up and down buttons.
```
or similar.

KoBeWi commented on editor/themes/editor_theme_manager.cpp:
```diff
@@ -1858,6 +1894,10 @@ void EditorThemeManager::_populate_editor_styles(const Ref<EditorTheme> &p_theme
 		editor_spin_label_bg->set_border_width_all(0);
 		p_theme->set_stylebox("label_bg", "EditorSpinSlider", editor_spin_label_bg);
 
+		// Should be changed to using separate arrows like on SpinBox; planned for a separate PR.
```
Add explicit "TODO" if changes are planned.

Mickeon commented on doc/classes/SpinBox.xml:
```diff
@@ -71,8 +71,98 @@
+			Vertical separation between the up and down buttons.
+		</theme_item>
+		<theme_item name="buttons_width" data_type="constant" type="int" default="16">
+			Width of the up and down buttons. If smaller than any icon set on the buttons, the said icon may overlap neighboring elements, unless [theme_item set_min_buttons_width_from_icons] is enabled.
```
```suggestion
			Width of the up and down buttons. If smaller than any icon set on the buttons, the respective icon may overlap neighboring elements, unless [theme_item set_min_buttons_width_from_icons] is different than [code]0[/code].
```

Mickeon commented on doc/classes/SpinBox.xml:
```diff
@@ -71,8 +71,98 @@
+		<theme_item name="field_and_buttons_separation" data_type="constant" type="int" default="2">
+			Width of the horizontal separation between the text input field ([LineEdit]) and the buttons.
+		</theme_item>
+		<theme_item name="set_min_buttons_width_from_icons" data_type="constant" type="int" default="1">
```
I have to say that that a property name prefixed with `set_` is rather unorthodox. I have no suggestions at the moment but it doesn't feel right.

davthedev replied:
Would something like "use_icon_width_as_min_button_width" work out better?

Sure, "set" and "get" may seem odd here with a programmer mindset as this is not a proper setter/getter in OOP sense. More a contraction of "Set the minimum button width according to the icon width". I thought about it more from a user standpoint that may be not biased with programming terms.

Mickeon replied:
Yes, I personally find that name to be much more suitable. Suggestions are still welcome if there's a briefer name.

Mickeon commented on doc/classes/SpinBox.xml:
```diff
@@ -71,8 +71,98 @@
+			Single up/down graphic for both up/down buttons. It is displayed in the middle of the buttons and does not change upon interaction. It is recommended to use individual [theme_item up] and [theme_item down] graphics for better usability. This can also be used as additional decoration between the two buttons.
+		</theme_item>
+		<theme_item name="down_background" data_type="style" type="StyleBox">
+			Background graphic of the down button.
```
All of these "graphic" below could be replaced by "style" (I've never seen these being referred as "graphic") 
```suggestion
			Background style of the down button.
```

davthedev replied:
See my comment on July 21 above for elaboration on how it works. Seems I cannot easily make a link to it.

davthedev replied:
Regarding set_min_buttons_width_from_icons, see my comment above on Jul 21 for explanation. Seems tricky to make a direct link to it.

Mickeon reviewed (commented):
Documentation is just fine now. My opinion on `set_min_buttons_width_from_icons` still remains but it shouldn't be an absolute blocker.

akien-mga commented:
Thanks!
